signer/threshold_signer_soft.go

```diff
@@ -0,0 +1,354 @@
+package signer
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+
+	tmCryptoEd25519 "github.com/tendermint/tendermint/crypto/ed25519"
+	tmJson "github.com/tendermint/tendermint/libs/json"
+	"gitlab.com/unit410/edwards25519"
+	tsed25519 "gitlab.com/unit410/threshold-ed25519/pkg"
+)
+
+// ThresholdSignerSoft implements the interface and signs the message for each local signer.
+// ThresholdSignerSoft is the implementation of a soft sign signer at the local level.
+type ThresholdSignerSoft struct {
+	PubKeyBytes []byte
+	Key         CosignerKey
+	// Total signers
+	Total     uint8
+	Threshold uint8
+	// Height, Round, Step, Timestamp --> metadata
+	HrsMeta map[HRSTKey]HrsMetadata
+}
+
+// NewThresholdSignerSoft constructs a ThresholdSigner
+// that signs using the local key share file.
+func NewThresholdSignerSoft(
+	key CosignerKey,
+	threshold, total uint8,
+) ThresholdSigner {
+	softSigner := &ThresholdSignerSoft{
+		Key:       key,
+		HrsMeta:   make(map[HRSTKey]HrsMetadata),
+		Total:     total,
+		Threshold: threshold,
+	}
+
+	// cache the public key bytes for signing operations
+	switch ed25519Key := softSigner.Key.PubKey.(type) {
+	case tmCryptoEd25519.PubKey:
+		softSigner.PubKeyBytes = make([]byte, len(ed25519Key))
+		copy(softSigner.PubKeyBytes, ed25519Key[:])
+	default:
+		panic("Not an ed25519 public key")
```

Comment:
This panic does not show the underlying type. 

If panicking anyway, then it's better to not switch on the type. Instead, `ed25519Key := softSigner.Key.PubKey.(tmcryptoed25519.PubKey)`.

Reply:
This is the current implementation as well! 
I think this is just to "assert" that the PubKey is a ed25519 key.